rs/nns/integration_tests/src/cycles_minting_canister.rs

```diff
@@ -778,6 +784,251 @@ fn test_cmc_cycles_create_with_settings() {
+            subaccount: Some(Subaccount::from(&TEST_USER1_PRINCIPAL.clone()).0),
+        },
+        amount: Nat::from(10 * E8),
+        fee: Some(Nat::from(10_000_u64)),
+
+        // Here, the "bomb is being planted".
+        memo: Some(icrc1::transfer::Memo(ByteBuf::from(garbage_memo))),
+
+        from_subaccount: None, // source from USER1's the default subaccount
+        created_at_time: None,
+    };
+    let create_canister_block_index = icrc1_transfer(
+        &state_machine,
+        LEDGER_CANISTER_ID,
+        *TEST_USER1_PRINCIPAL,
+        transfer_arg.clone(),
+    )
+    .expect("transfer failed");
+    assert_balance(90, 1, "ICP sent to CMC");
+
+    // This is to make it so that CMC has more ICP. That way, when we later try
+    // to duplicate the automatic refund, we can verify that CMC refrains from
+    // sending back more ICP.
+    let transfer_arg = icrc1::transfer::TransferArg {
+        amount: Nat::from(50 * E8),
+        ..transfer_arg
+    };
+    let _red_herring = icrc1_transfer(
+        &state_machine,
+        LEDGER_CANISTER_ID,
+        *TEST_USER1_PRINCIPAL,
+        transfer_arg,
+    )
+    .expect("transfer failed");
+    // Balance has gone down by 10 + 10 (plus fees). Later, we will see balance
+    // go back up, but only by 10 (minus fee).
+    assert_balance(40, 2, "additional red herring ICP sent to CMC");
+
+    // Step 2.2: Ask CMC to create a canister using the ICP that was just sent
+    // to it (from USER1). This is where it is noticed (by CMC) that USER1 did
+    // something wrong. Many requests are sent concurrently to verify that
+    // journaling/locking/deduplication works.
+    #[allow(deprecated)]
+    let notify_create_canister = Encode!(&NotifyCreateCanister {
+        block_index: create_canister_block_index,
+        controller: *TEST_USER1_PRINCIPAL,
+
+        // Nothing fancy.
+        subnet_type: None,
+        subnet_selection: None,
+        settings: None,
+    })
+    .unwrap();
+    let results = (0..100)
+        .map(|_i| {
+            state_machine.send_ingress_safe(
+                *TEST_USER1_PRINCIPAL,
+                CYCLES_MINTING_CANISTER_ID,
+                "notify_create_canister",
+                notify_create_canister.clone(),
+            )
+            .unwrap()
+        })
+        // It might seem silly to call collect, and then immediately after that
+        // call into_iter, but this is to ensure that await_ingress is not
+        // called until after ALL send_ingress_safe calls are made.
+        .collect::<Vec<_>>()
+        .into_iter()
+        .map(|message_id| {
+            let result = state_machine.await_ingress(message_id, 500).unwrap();
+
+            let result = match result {
+                WasmResult::Reply(ok) => ok,
+                _ => panic!("{:?}", result),
+            };
+
+            Decode!(&result, Result<CanisterId, NotifyError>).unwrap()
+        })
+        .collect::<Vec<_>>();
```

Comment:
Here is how I am trying to make concurrent calls (to CMC's notify_create_canister method).

Later, on line 975, it seems like there is no interleaving. If there were interleaving, I would expect to hit line 954, which would cause err to be of length < concurrency - 1, but it err.len() is always concurrency - 1, even when I crank concurrency all the way up to 10_000.